core/safemath/safe_math.go

```diff
@@ -0,0 +1,170 @@
+	// If the result should be negative, but the sign bit is unset, an overflow has occurred.
+	//
+	// The following examples use uint16/int8 as a more easily understandable example.
+	//
+	// *Positive Overflow*
+	//
+	// As a positive overflow example, take int8(66) * 2.
+	// The result as a uint16 will be 132, or in binary: 0000 0000 1000 0100.
+	// Interpreted as an int8 the high part is cut off and so the sign bit (most significant bit) is set,
+	// but we expected the result to be positive, so we detect this as an overflow.
+	//
+	// *Negative Overflow*
+	//
+	// As a negative overflow example, take int8(-66) * 2.
+	// The result as a uint16 will 132, or in binary 0000 0000 1000 0100, since we calculate 66*2 without the sign.
+	// Interpreted as an int8 the high part is cut off and the result multiplied with -1, to account for the
+	// expected sign of the result, so we get 0111 1100.
+	// The sign bit (most significant bit) is unset, but we expected the result to be negative,
+	// so we detect this as an overflow.
+	//
+	// *Overflowing 8-bit space*
+	//
+	// As a larger negative overflow example using 8-bit integers, take int8(-128) * 32.
+	// The multiplication would take place using 16-bits, so the result as a uint16 will be
+	// binary 0001 0000 0000 0000, since we calculate 128*32 without the sign.
+	// Interpreted as an int8 and multiplied with -1, to account for the expected sign of the result, we simply get
+	// 0000 0000, since the least significant byte of the result is 0.
+	// However, since the most significant byte of the result is non-zero, we detect the overflow.
+
+	xNegative := x < 0
+	yNegative := y < 0
+	xPositive := x > 0
+	yPositive := y > 0
+	resultIsPositive := true
+	var resultSign int64 = 1
+
+	// Store the sign of the resulting computation and negate any negative integers,
+	// so we can correctly convert to uint64.
+	if xNegative {
+		if yPositive {
+			resultIsPositive = false
+			resultSign = -1
+		}
+		x = -x
+	}
+
+	if yNegative {
+		if xPositive {
+			resultIsPositive = false
+			resultSign = -1
+		}
+		y = -y
+	}
+
+	// Execute the multiplication in 128-bit space using unsigned integers.
+	hi, lo := bits.Mul64(uint64(x), uint64(y))
+	// Interpret the lo result as an int64, then correct for the expected sign.
+	loSigned := int64(lo) * resultSign
+
+	// If the computation overflowed a uint64, it would also overflow an int64.
+	if hi != 0 {
+		return 0, ierrors.Wrapf(ErrIntegerOverflow, "%d and %d", x, y)
+	}
+
+	// Extract the most significant bit, signaling if the number is negative (1) or not (0).
+	signBitSet := ((loSigned >> 63) & 1) == 1
+
+	// If the result is expected to be positive but the sign bit is set, it's an overflow.
+	if resultIsPositive && signBitSet {
+		return 0, ierrors.Wrapf(ErrIntegerOverflow, "%d and %d", x, y)
+	}
+
+	// If the result is expected to be negative but the sign bit is not set, it's an underflow.
+	if !resultIsPositive && !signBitSet {
```

Comment:
why not one if-else for the resultIsPositive condition, and the other conditions in a second if inside the both cases? This way we don't need to check twice.